### PR TITLE
MAT-5323 terminology validation on Back End

### PR DESCRIPTION
**Pull request by adongare**
## MADiE PR

Jira Ticket: [MAT-5323](https://jira.cms.gov/browse/MAT-5323)
(Optional) Related Tickets:

### Summary
- moved createMeasure logic from controller to service
- translate cql if it is provided as part of create request payload and get elm
- validate the terminology
  
### All Submissions
* [x] This PR has the JIRA linked.
* [x] Required tests are included.
* [x] No extemporaneous files are included (i.e Complied files or testing results).
* [x] This PR is merging into the **correct branch**.
* [ ] All Documentation needed for this PR is Complete (or noted in a TODO or other Ticket).
* [ ] Any breaking changes or failing automations are noted by placing a comment on this PR.

### DevSecOps
If there is a question if this PR has a security or infrastructure impact, please contact the Security or DevOps engineer assigned to this project to discuss it further.

* [ ] This PR has NO significant security impact (i.e Changing auth methods, Adding a new user type, Adding a required but vulnerable package).

### Reviewers
By Approving this PR you are attesting to the following:

*  Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose.
*  The tests appropriately test the new code, including edge cases.
*  If you have any concerns they are brought up either to the developer assigned, security engineer, or leads.

### CODE REVIEW

File: src/main/java/cms/gov/madie/measure/config/TerminologyServiceConfig.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+  @Value("${madie.terminology.service.fetch-value-sets}")
+  private String valueSetsUrl;
+
+  @Value("${madie.terminology.service.validate-codes}")
+  private String validateCodeUrl;
+
+  @Bean
+  public RestTemplate terminologyRestTemplate() {
+    return new RestTemplate();
```
Comment (adongare) [inline src/main/java/cms/gov/madie/measure/config/TerminologyServiceConfig.java:24]: I'm seeing 3 or 4 rest templates being registered with spring context. Should I create a task to refactor this to have one common bean?

File: src/main/java/cms/gov/madie/measure/services/MeasureService.java [hunk 2]
```diff
@@ -36,6 +44,57 @@ public class MeasureService {
...
   private final MeasureUtil measureUtil;
+  private final ActionLogService actionLogService;
+  private final TerminologyValidationService terminologyValidationService;
+
+  public Measure createMeasure(Measure measure, final String username, String accessToken) {
+    log.info("User [{}] is attempting to create a new measure", username);
+    checkDuplicateCqlLibraryName(measure.getCqlLibraryName());
+    validateMeasurementPeriod(
+        measure.getMeasurementPeriodStart(), measure.getMeasurementPeriodEnd());
+    Measure measureCopy = measure;
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/services/MeasureService.java]: This does not create copy. It creates another pointer targeting the same object. 

I'm not sure a copy is needed, either.

Comment (adongare) [reply src/main/java/cms/gov/madie/measure/services/MeasureService.java]: a copy was not needed but codacy was complaining about reassigning method parameters in try-catch bock(next couple of lines to translate cql). So, just made up something. `Measure measureCopy = measure.toBuilder().build();` would have been better but facing some issues with test cases. The builder creates a singleton collection for errors object which can't be modified and then assertions fails.

Comment (adongare) [reply src/main/java/cms/gov/madie/measure/services/MeasureService.java]: okay, ignore that. I mocked data incorrectly. Updated it. Thanks Joe!

File: src/main/java/cms/gov/madie/measure/services/MeasureService.java [hunk 3]
```diff
@@ -36,6 +44,57 @@ public class MeasureService {
...
+
+  public Measure createMeasure(Measure measure, final String username, String accessToken) {
+    log.info("User [{}] is attempting to create a new measure", username);
+    checkDuplicateCqlLibraryName(measure.getCqlLibraryName());
+    validateMeasurementPeriod(
+        measure.getMeasurementPeriodStart(), measure.getMeasurementPeriodEnd());
+    Measure measureCopy = measure.toBuilder().build();
+    Set<MeasureErrorType> errorTypes = new HashSet<>();
+    try {
+      measureCopy = updateElm(measure, accessToken);
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/services/MeasureService.java]: Since `updateElm` may return the passed in `measure`, need to pass `measureCopy` to prevent it from being pointed to the original `measure`.

File: src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    }
+  }
+
+  public List<String> getValueSetOids(String elm) {
+    var library = getLibrary(elm);
+    var valueSets = (Map<String, Object>) library.get("valueSets");
+    if (valueSets == null) {
+      return null;
+    }
+    var valueSetDefs = (List<Map<String, String>>) valueSets.get("def");
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java:68]: I'm wondering if using a strict type instead of type casting would be preferred. I'll have to look it up.

File: src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        validatedCodes.stream().filter(cqlCode -> !cqlCode.isValid()).findFirst().orElse(null);
+    if (invalidCode != null) {
+      log.error("Invalid cql codes: " + invalidCode.getCodeId());
+      throw new InvalidTerminologyException("CQL Code", invalidCode.getCodeId());
+    }
+  }
+
+  public List<String> getValueSetOids(String elm) {
+    var library = getLibrary(elm);
+    var valueSets = (Map<String, Object>) library.get("valueSets");
```
Comment (jkotanchik-SB) [inline src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java:64]: I'm wondering if using a strict type instead of type casting would be preferred. I'll have to look it up.

Comment (jkotanchik-SB) [reply src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java:64]: Oh, this whole call stack is weird. I don't have any direct feedback atm, and will need to look into this further.

**Review by jkotanchik-SB — DISMISSED** [review]

File: src/main/java/cms/gov/madie/measure/services/MeasureService.java (hunk 3 above)
Comment (adongare) [reply src/main/java/cms/gov/madie/measure/services/MeasureService.java]: Joe with eagle's eyes!

File: src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java (hunk 5 above)
Comment (adongare) [reply src/main/java/cms/gov/madie/measure/services/TerminologyValidationService.java:64]: we might be able to get strict typing but types need to be included(translator). Another thing we could do is to move this to the translator. validate terminology and cql both

**Review by jkotanchik-SB — APPROVED** [review]
